### PR TITLE
Marketplace draft

### PR DESCRIPTION
**Pull request by joshuaschu**
I implemented mock UIs for the entry page and the search result page of the marketplace.
The detail view for a specific widget is nearly completely implemented. The only missing feature is the similar tab.

### CODE REVIEW

File: de.dashup.application/src/main/webapp/views/marketplace.jsp [hunk 1]
```diff
@@ -7,17 +9,136 @@
...
+                    <a href="${pageContext.request.contextPath}/" class="breadcrumb">dashup</a>
+                    <a href="#" class="breadcrumb"><fmt:message key="i18n.marketplace"/></a>
+                </div>
+            </div>
+        </nav>
+        <form action="${fn:escapeXml(pageContext.request.contextPath)}/marketplace/search" method="post">
+            <div class="row" style="margin-top: 25px">
+                <div class="col m4 offset-m4 s8 offset-s1">
+                    <div class="input-field">
+                        <input id="text-field-login-email" name="email" type="text" class="validate">
```
Comment (fidsusj) [inline de.dashup.application/src/main/webapp/views/marketplace.jsp]: guess you should change the id and for attribute here

File: de.dashup.application/src/main/webapp/views/marketplace.jsp [hunk 2]
```diff
@@ -7,17 +9,136 @@
...
+            <div class="col m6 offset-m3 s10 offset-s1">
+                <h3><fmt:message key="i18n.mostPopular"/></h3>
+                <hr/>
+            </div>
+        </div>
+        <div class="row">
+            <div class="col s12 m3 offset-m3">
+                <div class="card horizontal">
+                    <div class="card-image">
+                        <img src="https://lorempixel.com/100/190/nature/6">
```
Comment (fidsusj) [inline de.dashup.application/src/main/webapp/views/marketplace.jsp]: I should ask @raphaelmue whether he could add a feature to select an icon for the widget before publishing

File: de.dashup.application/src/main/webapp/views/marketplaceSearchResult.jsp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                <div class="col s12">
+                    <a href="${pageContext.request.contextPath}/" class="breadcrumb">dashup</a>
+                    <a href="#" class="breadcrumb"><fmt:message key="i18n.marketplace"/></a>
+                </div>
+            </div>
+        </nav>
+        <div class="row" style="margin-top: 25px">
+            <div class="col m4 offset-m4 s8 offset-s1">
+                <div class="input-field">
+                    <input id="text-field-login-email" name="email" type="text" class="validate">
```
Comment (fidsusj) [inline de.dashup.application/src/main/webapp/views/marketplaceSearchResult.jsp]: here once again the issue of the id attribute

File: de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java [hunk 4]
```diff
@@ -107,13 +107,117 @@ public Widget getPanelById(int id) throws SQLException {
...
-        List<? extends DatabaseObject> result = this.database.getObject(Database.Table.PANELS, Widget.class, whereParameters);
+        List<? extends DatabaseObject> result = this.database.getObject(Database.Table.PANELS, DatabaseWidget.class, whereParameters);
         if (result != null && result.size() == 1) {
             return (Widget) new Widget().fromDatabaseObject(result.get(0));
         }
         return null;
     }
 
+    //-------------- Marketplace --------------\\
+    public ArrayList<String> getTagsByPanelId(int panelId) throws SQLException {
```
Comment (fidsusj) [inline de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java]: Rename this as below into ...WidgetID

**Comment by joshuaschu** [comment]
> * When trying to add a panel to the dashboard or submitting a rating I get the following error message: Problems calling function [fn:escapeXml], so probably some jsp escaping issue? (only for wheather and share banner, the other two are working, at least for adding to the dashboard)
Fixed with 18da084 ... there where some missing getters in the ratings class so no info in the frontend...
 
> * when being on the entry page of the marketplace, making the screen size smaller causes the descriptive text to overflow the card boundaries (and removing the "show more" option)

I need to rework this when finally creating the marketplace front site....

> * please align the widget icon to the left border of the page (the same spacing as the overview menu) or at least add some spacing between the icon and the widget name in the detail view (especially when making the screen smaller, otherwise it will overlay)
Increased offset between icon and text a little bit, but I dont really know how to fix this completely cause we can use the m and s classes but the change from m to s is a little bit to late so there is this overlay... we can also consider making the icon smaller but i think this is not that great...  

> * when the marketplace is merged, you will need to add "category" in the additional information section
Cant do this now as you mentioned.

> * widgets should not be added twice to the dashboard
Why not? Two weather widgets are nice i think... especially when you are able to enter two citys via properties...

**Comment by joshuaschu** [comment]
[dashup_dev (3).zip](https://github.com/raphaelmue/dashup/files/3197252/dashup_dev.3.zip)

This is the db scheme i am using for this branch.

File: de.dashup.application/src/main/resources/i18n_de.properties [hunk 5]
```diff
@@ -66,6 +66,16 @@ i18n.save=Speichern
 i18n.undo=R�ckg�ngig
 i18n.successChangedLayout=Die Darstellung wurde erfolgreich ge�ndert!
 i18n.passwordLength=Das Passwort muss eine Mindestl�nge von 8 Zeichen haben.
+i18n.enterSearchterm=Hier den Suchbegriff eingeben...
+i18n.featured=Vorgestellt
+i18n.go=Suchen
+i18n.mostPopular=Am beliebtesten
+i18n.bestRating=Am besten bewertet
+i18n.add=Hinzuf�gen
+i18n.dateOfPublication=Ver�ffentlichungsdatum:
```
Comment (raphaelmue) [inline de.dashup.application/src/main/resources/i18n_de.properties]: Rather "Erscheinungsdatum"

File: de.dashup.application/src/main/java/de/dashup/application/controllers/MarketplaceController.java [hunk 6]
```diff
@@ -15,8 +21,62 @@
...
+                            @RequestParam("sectionId") int sectionId,
+                            @RequestParam("widgetSize") String widgetSize) throws SQLException {
+
+        User user = LocalStorage.getInstance().getUser(request, token);
+        if (user != null) {
+            boolean success = DashupService.getInstance().addWidget(user, widgetId, sectionId, widgetSize);
+            if (success) {
+                return "redirect:/marketplace/detailView/" + widgetId + "#addedWidget";
+            } else {
+                return "redirect:/marketplace/detailView/" + widgetId + "#faieldToAddWidget";
```
[the remark this answers is not in the document]
Comment (joshuaschu) [reply de.dashup.application/src/main/java/de/dashup/application/controllers/MarketplaceController.java]: done

File: de.dashup.application/src/main/webapp/views/marketplace.jsp [hunk 7]
```diff
@@ -7,17 +9,136 @@
...
+                    </div>
+                    <div class="card-stacked">
+                        <div class="card-content">
+                            <p>I am a very simple card. I am good at containing small bits of information.</p>
```
[the remark this answers is not in the document]
Comment (joshuaschu) [reply de.dashup.application/src/main/webapp/views/marketplace.jsp]: I think this belongs to the implementation of frontpage of marketplace. I can do this now but i thought that we want to merge before everything is finished, wont we?

File: de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java [hunk 8]
```diff
@@ -107,13 +107,117 @@ public Widget getPanelById(int id) throws SQLException {
         Map<String, Object> whereParameters = new HashMap<>();
         whereParameters.put("id", id);
 
-        List<? extends DatabaseObject> result = this.database.getObject(Database.Table.PANELS, Widget.class, whereParameters);
+        List<? extends DatabaseObject> result = this.database.getObject(Database.Table.PANELS, DatabaseWidget.class, whereParameters);
         if (result != null && result.size() == 1) {
             return (Widget) new Widget().fromDatabaseObject(result.get(0));
```
[the remark this answers is not in the document]
Comment (joshuaschu) [reply de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java]: wont fix as discussed...

File: de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java [hunk 9]
```diff
@@ -107,13 +107,117 @@ public Widget getPanelById(int id) throws SQLException {
...
+            for (DatabaseObject databaseObject : result) {
+                Widget widget = (Widget) new Widget().fromDatabaseObject(databaseObject);
+                widget.setShortDescription(this.shortenShortDescOfPanel(widget.getShortDescription()));
+                returningValue.add(widget);
+            }
+        }
+        return returningValue;
+    }
+
+    private String shortenShortDescOfPanel(String shortDescr) {
```
[the remark this answers is not in the document]
Comment (joshuaschu) [reply de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java]: I think this belongs to the implementation of frontpage too. So yes ill use this class when doing the frontpage.

File: de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java [hunk 10]
```diff
@@ -107,13 +107,117 @@ public Widget getPanelById(int id) throws SQLException {
...
+    public boolean addRating(User user, String title, String text, int rating, int widgetId) {
+        Map<String, Object> insertValue = new HashMap<>();
+        insertValue.put("user_id", user.getId());
+        insertValue.put("panel_id", widgetId);
+        insertValue.put("rating", rating);
+        insertValue.put("title", title);
+        insertValue.put("text", text);
+        insertValue.put("changed_on", new Date());
+        try {
+            database.insert(Database.Table.RATINGS, insertValue);
```
[the remark this answers is not in the document]
Comment (joshuaschu) [reply de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java]: I think this failure should not cause the app to fail and show a error page... its a little failure and i think it doesn't really matter if this doesn't work. So i prevented the exception.

File: de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java [hunk 11]
```diff
@@ -107,13 +107,117 @@ public Widget getPanelById(int id) throws SQLException {
...
+        } catch (SQLException e) {
+            return false;
+        }
+        return true;
+    }
+
+    public boolean addWidget(User user, int widgetId, int sectionId, String widgetSize) throws SQLException {
```
[the remark this answers is not in the document]
Comment (joshuaschu) [reply de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java]: renamed to addWidgetToPersonalDashup

File: de.dashup.shared/src/main/java/de/dashup/shared/Widget.java [hunk 12]
```diff
@@ -92,9 +96,9 @@ public String getCode(Size size) {
 
     public String getCodeWithWrapper() {
         return "<div class=\"card col " + this.size.getStyleClass() + "\">" +
-                    "<div class=\"card-content\">" +
-                        this.getCode() +
-                    "</div>" +
+                "<div class=\"card-content\">" +
+                this.getCode() +
+                "</div>" +
```
[the remark this answers is not in the document]
Comment (joshuaschu) [reply de.dashup.shared/src/main/java/de/dashup/shared/Widget.java]: Not my fault... so i dont know ill leave it like it is

**Comment by joshuaschu** [comment]
Here is the correct DB scheme with the needed data:
[MarketplaceDB.zip](https://github.com/raphaelmue/dashup/files/3215470/MarketplaceDB.zip)

File: de.dashup.application/src/main/java/de/dashup/application/controllers/MarketplaceController.java [hunk 13]
```diff
@@ -19,7 +22,11 @@
     @RequestMapping("/")
     public String marketplace(@CookieValue(name = "token", required = false) String token, Model model, HttpServletRequest request) throws SQLException {
         return ControllerHelper.defaultMapping(token, request, model, "marketplace", user -> {
-            // tbd
+            model.addAttribute("carouselWidget",DashupService.getInstance().getPanelById(1));
+            int[] featuredWidgets = {1,2,3,4};
```
Comment (fidsusj) [inline de.dashup.application/src/main/java/de/dashup/application/controllers/MarketplaceController.java]: We should change this to four random numbers from all widgets available lateron

File: de.dashup.application/src/main/webapp/views/marketplace.jsp [hunk 14]
```diff
@@ -7,17 +9,166 @@
...
+                                <div class="row" style="margin-left: 10px; margin-right: 10px;">
+                                    <p>
+                                        ${fn:escapeXml(mapEntry.key.shortDescription)}
+                                    </p>
+                                </div>
+                                <div class="row" style="margin: 0 auto; width: 164px">
+                                    <div class="col">
+                                        <div class="star-rating" style="font-size: 25px; margin-top: 15px; color: var(--color-dark-gray);">
+                                            <div class="back-stars">
+                                                <i class="fa fa-star" aria-hidden="true"></i>
```
Comment (fidsusj) [inline de.dashup.application/src/main/webapp/views/marketplace.jsp]: You could use the jsp for loop here

File: de.dashup.application/src/main/webapp/views/marketplace.jsp [hunk 15]
```diff
@@ -7,17 +9,166 @@
...
+                                                <i class="fa fa-star" aria-hidden="true"></i>
+                                                <i class="fa fa-star" aria-hidden="true"></i>
+                                                <i class="fa fa-star" aria-hidden="true"></i>
+                                                <i class="fa fa-star" aria-hidden="true"></i>
+                                                <div class="front-stars"
+                                                     style="width:  ${fn:escapeXml(mapEntry.key.averageRating)}%;color: var(--color-background);">
+                                                    <i class="fa fa-star" aria-hidden="true"></i>
```
Comment (fidsusj) [inline de.dashup.application/src/main/webapp/views/marketplace.jsp]: here as well

File: de.dashup.application/src/main/webapp/views/marketplace.jsp [hunk 16]
```diff
@@ -7,17 +9,166 @@
...
+        <div class="row">
+            <div class="col m6 offset-m3 s10 offset-s1">
+                <h3><fmt:message key="i18n.mostPopular"/></h3>
+                <hr/>
+            </div>
+        </div>
+        <div class="row">
+            <div class="col s12 m6 offset-m3">
+                <div class="row">
+                    <c:forEach items="${mostDownloaded}" var="panel">
```
Comment (fidsusj) [inline de.dashup.application/src/main/webapp/views/marketplace.jsp]: once again you could use a for loop over mostDownloaded and bestRated

File: de.dashup.application/src/main/webapp/views/marketplaceSearchResult.jsp [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+                    <li class="collection-item" id="weather-panel">
+                        <div class="row">
+                            <div class="col m3 s3">
+                                <i class="fas fa-cloud fa-7x"></i>
+                            </div>
+                            <div class="col m4 s4">
+                                <h5>Weather Panel</h5>
+                                <div>This is the short description of this panel</div>
+                                <div style="margin-top: 5px">
+                                    <i class="fas fa-star"></i>
```
Comment (fidsusj) [inline de.dashup.application/src/main/webapp/views/marketplaceSearchResult.jsp]: here as well

File: de.dashup.model/src/main/java/de/dashup/model/db/Database.java [hunk 18]
```diff
@@ -201,23 +214,38 @@ public JSONArray get(Table table, Map<String, Object> whereParameters) throws SQ
...
+     * @param whereParameters parameters that are used in the where clause to select data
+     * @return result of the database query
+     */
+    public JSONArray get(Table table, Table joinOn, Map<String, String> onParameters, Map<String, Object> whereParameters, String orderByClause) throws SQLException {
         PreparedStatement statement;
-        String query = "SELECT * FROM " + table.getTableName() + " INNER JOIN " + joinOn.getTableName() +
-                " ON ";
-        for (Map.Entry<String, Object> entry : onParameters.entrySet()) {
-            query += table.getTableName() + "." + entry.getKey() + " = " + joinOn.getTableName() + "." +entry.getValue();
+        StringBuilder query = new StringBuilder("SELECT * FROM " + table.getTableName() + " INNER JOIN " + joinOn.getTableName() +
```
Comment (fidsusj) [inline de.dashup.model/src/main/java/de/dashup/model/db/Database.java]: This is a potential SQL injection risk. Please use the prepared statement with the "?" as placeholders instead

File: de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java [hunk 19]
```diff
@@ -109,13 +109,161 @@ public Widget getPanelById(int id) throws SQLException {
...
+
+    public Map<Widget, Rating> getFeaturedWidgets(int[] widgetIds) throws SQLException {
+        Map<Widget, Rating> returningValue = new HashMap<>();
+        for (int widgetId : widgetIds) {
+            returningValue.put(this.getPanelById(widgetId), this.getTopRating(widgetId));
+        }
+        return returningValue;
+    }
+
+    private Rating getTopRating(int widgetId) throws SQLException {
```
Comment (fidsusj) [inline de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java]: Please use the SQL AVG Aggregation and don't take the best ranking for a widget, but the average one.

**Comment by joshuaschu** [comment]
> * Missing recently uploaded section on the start page.

Why should we do that... I mean it looks very bad when there are so many categories with all the same panels right? So I think the categories we have are enough.

> * Please adapt the color of the caruuuusel to the theme.

The color is adapted to the secondary color of the theme. If you dont like the color u need to change the themes.

> * Missing category detail on additional information
> * display publisher name next to rating on detail page (see mockup)

added in 891161b.

File: de.dashup.application/src/main/java/de/dashup/application/controllers/MarketplaceController.java (hunk 13 above)
Comment (joshuaschu) [reply de.dashup.application/src/main/java/de/dashup/application/controllers/MarketplaceController.java]: Yes lateron, but not now.... we agreed to hardcoded widgets in this section so I think u should not mention that here...

File: de.dashup.application/src/main/webapp/views/marketplace.jsp (hunk 14 above)
Comment (joshuaschu) [reply de.dashup.application/src/main/webapp/views/marketplace.jsp]: I can use the jsp loop to iterate over a list or map or something but not like `for(int i= 0;i<5;i++)`.
Feel free to inform yourself here: https://www.tutorialspoint.com/jsp/jsp_standard_tag_library.htm.
In addition to that I would not do a for loop here cause its just 2 lines shorter and not that readable.

File: de.dashup.application/src/main/webapp/views/marketplace.jsp (hunk 15 above)
Comment (joshuaschu) [reply de.dashup.application/src/main/webapp/views/marketplace.jsp]: See above

File: de.dashup.application/src/main/webapp/views/marketplace.jsp (hunk 16 above)
Comment (joshuaschu) [reply de.dashup.application/src/main/webapp/views/marketplace.jsp]: See line 103: there is the forEach loop??????????????
If u know a better way plz let me know....

File: de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java (hunk 19 above)
Comment (joshuaschu) [reply de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java]: On the carusel the rating is taken from the widget. But i need this function to get the featured comment so this is absolutely correct. We agreed on a hard coded carusel so here is it... on db we can create comments with rating 101 so they are used for the carusel...

File: de.dashup.model/src/main/java/de/dashup/model/db/Database.java (hunk 18 above)
Comment (joshuaschu) [reply de.dashup.model/src/main/java/de/dashup/model/db/Database.java]: I used common way to access the db... like it is in every other method in this class...
The prepared statement will take care of the where parameters (have a look at line 237). The table names and the on parameters CAN NOT be adapted by the user cause they are "hardcoded" in the dashup service which is in the backend and completely encapsulated and isolated from the User in front of the webpage. You can ask our special technological idontknowwhat @raphaelmue about this and he will confirm this!!!

**Comment by joshuaschu** [comment]
Hey @fidsusj,
in your last comment about the src code all the points contains some technical or programmatic mistakes. I would appreciate if you inform yourself before criticizing someones code you obviously do not understand.
Regards,
Joshua

File: de.dashup.application/src/main/webapp/views/marketplace.jsp (hunk 16 above)
Comment (fidsusj) [reply de.dashup.application/src/main/webapp/views/marketplace.jsp]: Yes, you made a JSLT for loop to generate the widgets into the sections mostDownloaded and bestRated. But as these sections are created the same way, you could have created a for loop over the creation of these sections as well. Besides JSLT you can use so called scriplets in JSP as well. They allow you to write normal Java Code in JSP Views, which is one reason, why they are called Java Server Pages. I pushed some sample coding commented at the concerned line, can you implement this real quick?

File: de.dashup.application/src/main/webapp/views/marketplaceSearchResult.jsp (hunk 17 above)
Comment (fidsusj) [reply de.dashup.application/src/main/webapp/views/marketplaceSearchResult.jsp]: I did this for you with scriptlets

File: de.dashup.model/src/main/java/de/dashup/model/db/Database.java (hunk 18 above)
Comment (fidsusj) [reply de.dashup.model/src/main/java/de/dashup/model/db/Database.java]: I see this is not your fault. I will need to talk to @raphaelmue as using this.getClause(whereParameters, "WHERE", " AND ") with a whereParameters map filled with user input can lead to an SQL injection. getClause, unfortunately, uses a StringBuilder inside,which is why user input will not be treated securly by the preparedStatement. Normally you would use preparedStatement.setString(1, <your string>) to treat user input securly.

File: de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java (hunk 19 above)
Comment (fidsusj) [reply de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java]: Ah, now I understand the use case of this method. Didn't know that this is used for the carousel, thought it is used for marketplace result list ^^. My fault!

File: de.dashup.application/src/main/java/de/dashup/application/controllers/MarketplaceController.java (hunk 13 above)
Comment (fidsusj) [reply de.dashup.application/src/main/java/de/dashup/application/controllers/MarketplaceController.java]: I implemented the logic for you using the RAND() function of maria db in combination with LIMIT. Now you can enhance the amout of widgets in the featured section by changing one number.

File: de.dashup.application/src/main/webapp/views/marketplace.jsp (hunk 14 above)
Comment (fidsusj) [reply de.dashup.application/src/main/webapp/views/marketplace.jsp]: Using JSLT you can only use forEach on pageContext objects. But you can as well use scriptlets, printing out markup to the JSP view or adding new pageContext objects. Did it for you.

Comment (fidsusj) [reply de.dashup.application/src/main/webapp/views/marketplace.jsp]: http://www.jsptut.com/Scriptlets.jsp
Here you go

File: de.dashup.application/src/main/webapp/views/marketplace.jsp (hunk 15 above)
Comment (fidsusj) [reply de.dashup.application/src/main/webapp/views/marketplace.jsp]: answered above

**Comment by fidsusj** [comment]
> in your last comment about the src code all the points contain some technical or programmatic mistakes. I would appreciate if you inform yourself before criticizing someone's code

First of all, I'm not criticizing your code, I am reviewing your code. If you don't appreciate the efforts I spent into reviewing your code, then please tell me, I won't take it personally. Secondly, I admit that you cannot know about JSP scriptlets and how they can make life easier. The only thing I didn't know was the use case of the getTopRating method, sorry for that!
And @raphaelmue and me will have to talk about the use of the preparedStatement again.

**Comment by Kucki99** [comment]
Hi Felix,
scriptlets are bad practice. Therefore I do not recommend to use them in your project.

Kind Regards,
Robin

**Pull request merged**